py/dml/codegen.py

```diff
@@ -1573,10 +1573,19 @@ def stmt_log(stmt, location, scope):
         if location.method():
             key = mkLit(site, '(uint64)NULL', TInt(64, False))
         else:
-            # Normally, we are not allowed to cast "this.trait" to an uint,
-            # but since we _internally_ know it is a pointer this is justified
-            key = mkLit(site, "(uint64)(%s.trait)" % lookup_var(
-                site, scope, "this").read(), TInt(64, False))
+            # Acquire a key based on trait _identity_t
+            identity = TraitObjIdentity(site, lookup_var(site, scope, "this"))
+            # identity is opaque, so we need to cheat a little bit
+            top_bits = mkShL(site,
```

Comment:
Maybe the test should also cover the array logic? i.e. extend g to `g[i<2]` and check that it will log once on `[0]` and once on `[1]`.

Reply:
reasonable. will do

Reply:
Done